test/cluster/mzcompose.py

```diff
@@ -1077,3 +1078,116 @@ def fetch_reconciliation_metrics() -> Tuple[int, int]:
     # TODO(#17594): Flip these once the bug is fixed.
     assert reused == 0
     assert replaced == 4
+
+
+def workflow_test_mz_subscribes(c: Composition) -> None:
+    """
+    Test that in-progress subscribes are reflected in mz_subscribes.
+    """
+
+    c.down(destroy_volumes=True)
+
+    c.up("testdrive", persistent=True)
+    c.up("materialized", "clusterd1", "clusterd2")
+
+    c.sql(
+        """
+        CREATE CLUSTER cluster1 REPLICAS (r (
+                STORAGECTL ADDRESSES ['clusterd1:2100'],
+                STORAGE ADDRESSES ['clusterd1:2103'],
+                COMPUTECTL ADDRESSES ['clusterd1:2101'],
+                COMPUTE ADDRESSES ['clusterd1:2102'],
+                WORKERS 1
+        ));
+        CREATE CLUSTER cluster2 REPLICAS (r (
+                STORAGECTL ADDRESSES ['clusterd2:2100'],
+                STORAGE ADDRESSES ['clusterd2:2103'],
+                COMPUTECTL ADDRESSES ['clusterd2:2101'],
+                COMPUTE ADDRESSES ['clusterd2:2102'],
+                WORKERS 1
+        ));
+        CREATE TABLE t (a int);
+        """
+    )
+
+    def check_mz_subscribes(expected: Tuple) -> None:
+        """
+        Check that the expected subscribes exist in mz_subscribes.
+        We identify subscribes by user and cluster only. We explicitly don't
+        check the `GlobalId`, as how that is allocated is an implementation
+        detail and might change in the future.
+        """
+        output = c.sql_query(
+            f"""
+            SELECT s.user, c.name
+            FROM mz_internal.mz_subscribes s
+              JOIN mz_clusters c ON (c.id = s.cluster_id)
+            ORDER BY s.started_at
+            """
+        )
+        assert output == expected, f"expected: {expected}, got: {output}"
+
+    def run_subscribe(id_: int, on_cluster: str) -> None:
+        """
+        Run a subscribe with on the given cluster.
+        The subscribe tries to fetch one record with a value of `id_` from
+        table `t`, which means it can be stopped by INSERTing that value.
+        """
+        c.sql(
+            f"""
+            SET cluster = {on_cluster};
+            BEGIN;
+            DECLARE c CURSOR FOR SUBSCRIBE (SELECT * FROM t WHERE a = {id_});
+            FETCH 1 c WITH (TIMEOUT = '10s');
+            ROLLBACK;
+            """
+        )
+
+    # Start first subscribe.
+    subscribe1 = Thread(target=run_subscribe, args=("1", "default"))
+    subscribe1.start()
+    time.sleep(1)
+
+    check_mz_subscribes((["materialize", "default"],))
+
+    # Start second subscribe.
+    subscribe2 = Thread(target=run_subscribe, args=("2", "cluster1"))
+    subscribe2.start()
+    time.sleep(1)
```

Comment:
I'm a little worried that these sleeps will cause flaky tests. Is there any retry mechanism for these tests? We could instead retry `check_mz_subscribes` until it passes.

Reply:
I tried to use testdrive to implement the checks. That would perform retries automatically. Other cluster tests use it, but for some reason I didn't get it to work due to docker-compose issues. I think I can try a bit harder though.